In [1]:
import numpy as np
from matplotlib import pyplot as plt
from sklearn import model_selection, datasets, naive_bayes

%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [7]:
digits = datasets.load_digits()
digits

{'DESCR': "Optical Recognition of Handwritten Digits Data Set\n===================================================\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 5620\n    :Number of Attributes: 64\n    :Attribute Information: 8x8 image of integer pixels in the range 0..16.\n    :Missing Attribute Values: None\n    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)\n    :Date: July; 1998\n\nThis is a copy of the test set of the UCI ML hand-written digits datasets\nhttp://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits\n\nThe data set contains images of hand-written digits: 10 classes where\neach class refers to a digit.\n\nPreprocessing programs made available by NIST were used to extract\nnormalized bitmaps of handwritten digits from a preprinted form. From a\ntotal of 43 people, 30 contributed to the training set and different 13\nto the test set. 32x32 bitmaps are divided into nonoverlapping blocks of\n4x4 and the number of on pixels ar

In [3]:
estimator = naive_bayes.BernoulliNB()

print (model_selection.cross_val_score(estimator, digits.data, digits.target).mean())

estimator = naive_bayes.MultinomialNB()

print (model_selection.cross_val_score(estimator, digits.data, digits.target).mean())

estimator = naive_bayes.GaussianNB()

print(model_selection.cross_val_score(estimator, digits.data, digits.target).mean())

0.825823650778
0.870877148974
0.818600380355


In [8]:
breast_cancer = datasets.load_breast_cancer()
breast_cancer

{'DESCR': 'Breast Cancer Wisconsin (Diagnostic) Database\n=============================================\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry \n        - fractal dimension ("coastline approximation" - 1)\n\n        The mean, standard error, and "worst" or largest (mean of the three\n        largest values) of these features were computed for each image,\n        resulting in 30 features.  For instance, field 3 i

In [5]:
estimator = naive_bayes.BernoulliNB()

print (model_selection.cross_val_score(estimator, breast_cancer.data, breast_cancer.target).mean())

estimator = naive_bayes.MultinomialNB()

print (model_selection.cross_val_score(estimator, breast_cancer.data, breast_cancer.target).mean())

estimator = naive_bayes.GaussianNB()

print(model_selection.cross_val_score(estimator, breast_cancer.data, breast_cancer.target).mean())

0.627420402859
0.894579040193
0.936749280609


Максимальное качество классификации на breast_cancer равно 0.93, на digits -- 0.87.

Вещественные признаки -- из второго набора. Здесь лучше всего сработал Гауссовский НБ.

Целые неотрицательные -- из первого набора. Здесь чемпион -- мультиномиальный НБ.

Таким образом, ответы на вопросы: нет нет да да.